### PR TITLE
fix update node condition

### LINKED ISSUE #65051: Update node condition in cloud node controller doesn't  work actually.
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
# > /kind feature

**What happened**:
In [cloud node controller(line 380)](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L380), the controller add NodeNetworkUnavailable to the node conditions, and use `cnc.kubeClient.CoreV1().Nodes().Update(curNode)` to update it. But node status can not be updated by `Nodes().Update()`, it doesn't work because [node update strategy(line 79)](https://github.com/kubernetes/kubernetes/blob/2fcac6abf27a3912a71f590300b72c2e48266120/pkg/registry/core/node/strategy.go#L79) replace the node status with old status. 
To fix this bug, we should use 'nodeutil.SetNodeCondition()' to update node status.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by metayd**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #65051

/area cloudprovider
/sig cluster-lifecycle

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by dbdd4us** [comment]
It seems `PatchStatus` method of  `FakeNodeHandler` is not implemented, so implement it.

File: pkg/controller/cloud/node_controller.go [hunk 1]
```diff
@@ -337,6 +337,21 @@ func (cnc *CloudNodeController) AddCloudNode(obj interface{}) {
 	}
 
 	err := clientretry.RetryOnConflict(UpdateNodeSpecBackoff, func() error {
+		// TODO(wlan0): Move this logic to the route controller using the node taint instead of condition
```
Comment (andrewsykim) [inline pkg/controller/cloud/node_controller.go]: What's the justification of moving this here earlier? Is it to set the node condition as early as possible?

cc @cheftako

Comment (dbdd4us) [reply pkg/controller/cloud/node_controller.go]: Yes, we must ensure that `NodeNetworkUnavailable` condition is added before we remove the cloud taint.

Btw, by moving these code ahead of the `curNode, err := cnc.kubeClient.CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})`, the curNode's would have `NodeNetworkUnavailable`  condition after we get it from apiserver by `cnc.kubeClient.CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})`.
If we don't move these code, when we update the node condition with `nodeutil.SetNodeCondition` methods, the curNode's status still doesn't have `NodeNetworkUnavailable`  condition, which is a little inconsistent, and we need to `Get` the node from apiserver again.

Comment (andrewsykim) [reply pkg/controller/cloud/node_controller.go]: Not sure I follow, the current implementation is 
```
curNode.Status.Conditions = append(node.Status.Conditions, v1.NodeCondition{
				Type:               v1.NodeNetworkUnavailable,
				Status:             v1.ConditionTrue,
				Reason:             "NoRouteCreated",
				Message:            "Node created without a route",
				LastTransitionTime: metav1.Now(),
			})
```
Which updates the condition on the node object in memory but does not apply any changes to the apiserver yet. The current implementation bundles all the necessary updates (removing taints, conditions, zones, instance types, etc) into 1 update call [here](https://github.com/kubernetes/kubernetes/pull/65052/files#diff-a1ff61328540d76c791d919f71da4ad5R409) so there is no need to call `nodeutil.SetNodeCondition` as far as I know.

Comment (jhorwit2) [reply pkg/controller/cloud/node_controller.go]: Is it OK to remove this condition before we know the network is actually ok? I think we need someone from GCP to review this as well.

Comment (dbdd4us) [reply pkg/controller/cloud/node_controller.go]: @andrewsykim 

The one update call doesn't update the node condition actually, the reason was describe here. #65051

Comment (andrewsykim) [reply pkg/controller/cloud/node_controller.go]: This is good to know, thanks :)

**Comment by andrewsykim** [comment]
cc @cheftako for GCE review

File: pkg/controller/cloud/node_controller.go (hunk 1 above)
Comment (cheftako) [reply pkg/controller/cloud/node_controller.go]: @dbdd4us, @andrewsykim 
I think it is correct to move the no route condition earlier. I'm slightly worried about the earlier error exit we are adding. We seem to have several such early error exits, so nothing in this block are guaranteed to happen. I don't think this is particular branch is actually exercised today. I should be testing this well before it is enabled on the main line. Given that I am ok with the change.

**Review by cheftako — APPROVED** [review]
/lgtm

**Comment by dbdd4us** [comment]
Sorry to bother, is there any update for this pr ?

**Comment by dbdd4us** [comment]
/assign @derekwaynecarr 

Sorry to bother, could you take a look at this pr? Thanks : )

**Comment by dbdd4us** [comment]
Any update?

**Review by janetkuo — APPROVED** [review]

**Pull request merged**